modules/seqimporter/src/main/scala/gem/seqimporter/SequenceImporter.scala

```diff
@@ -0,0 +1,154 @@
+          <params><param><value><string>{oid}</string></value></param></params>
+        </methodCall>
+      }
+
+    // Run the request on the ODB, which returns the sequence XML if all goes
+    // well wrapped in XML-RPC elements.
+    client.expect[Elem](req).map { elem =>
+      // Fish the return value out of the XML-RPC result.
+      val xmlTxt = (elem \\ "value").text
+
+      // Drop the preliminary cruft and then parse the sequence XML.  We're
+      // assuming that the observation was found in the ODB and all went well
+      // or else this will fail with an exception.  We could fish out XML-RPC
+      // failure results and return more accurate failure information if needed.
+      XML.loadString(xmlTxt.lines.drop(3).mkString("\n"))
+    }
+  }
+
+  /** Converts the OCS2 sequence XML into the new sequence model.  Each step
+    * includes the instrument configuration and its completion state.
+    */
+  private def parseSequenceXml(oid: Observation.Id, xml: Elem): List[Step[(InstrumentConfig, Boolean)]] = {
+    // Extracts the steps in the XML sequence to a simple list of Maps where
+    // each Map is all the keys and values that apply to the step.
+    def configMaps(xml: Elem): List[ConfigMap] = {
+      def extractSystem(configMap: ConfigMap, sys: Node): ConfigMap = {
+        val sysName = (sys \ "@name").text
+
+        (configMap /: (sys \ "param")) { (map, param) =>
+          val paramName  = (param \ "@name").text
+          val paramValue = (param \ "@value").text
+          map + (s"$sysName:$paramName" -> paramValue)
+        }
+      }
+
+      def extractStep(configMap: ConfigMap, step: Node): ConfigMap =
+        (configMap /: (step \ "system")) { extractSystem }
+
+      (xml \ "step").toList.scanLeft(EmptyConfigMap) { extractStep }.tail
+    }
+
+    configMaps(xml).map(ConfigReader.unsafeStep)
+  }
+
+  /** Fetches a Sequence from the ODB using its WDBA service, translating it
+    * into the new sequence model in memory.
+    */
+  def fetchSequence(oid: Observation.Id): Task[List[Step[(InstrumentConfig, Boolean)]]] =
+    fetchSequenceXml(oid).map { parseSequenceXml(oid, _) }
+
+  private val configureLogging: Task[Unit] =
+    Task.delay(List(
+      "edu.gemini.spModel.type.SpTypeUtil"
+    ).map(Logger.getLogger).foreach(_.setLevel(Level.OFF)))
+
+  // We need a minimal program/observation in order to insert steps.  The
+  // seqexec doesn't actually use these but they have to be there to satisfy
+  // database constraints.
+  private def createMinimalContext(oid: Observation.Id, s: List[Step[(InstrumentConfig, Boolean)]]): Task[Unit] = {
+    val i = s.headOption.map(_.instrument._1.instrument)
+    val p = Program[Nothing](oid.pid, "", Nil)
+    val o = Observation[Nothing](oid, "", i, Nil)
+
+    for {
+      _ <- ProgramDao.insert(p).onUniqueViolation(0.point[ConnectionIO]).transact(lxa)
+      _ <- ObservationDao.insert(o).onUniqueViolation(0.point[ConnectionIO]).transact(lxa)
+    } yield ()
+  }
```

Comment:
Hi, sorry to keep picking at this but I would prefer to see it as a single transaction so that if `ObservationDao.insert` fails for a reason other than unique violation the `ProgramDao.insert` will get rolled back as well. 

We can define a combinator that rolls back a single operation on unique violation

```scala
def ignoreUniqueViolation(fa: ConnectionIO[Int]): ConnectionIO[Int] =
  for {
    s <- HC.setSavepoint
    n <- fa.onUniqueViolation(HC.rollback(s).as(0))
    _ <- HC.releaseSavepoint(s)
  } yield n
```

which lets us write the insert as

```scala
val prog = 
  ignoreUniqueViolation(ProgramDao.insert(p)) *> 
  ignoreUniqueViolation(ObservationDao.insert(o))
  
prog.transact(xa)
```